### PR TITLE
RavenDB-16079 normalize new line characters when comparing maps or reduce

### PR DESCRIPTION
**Pull request by ppekrol**

### CODE REVIEW

File: src/Raven.Client/Util/NewLineInsensitiveStringComparer.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text.RegularExpressions;
+
+namespace Raven.Client.Util
+{
+    internal sealed class NewLineInsensitiveStringComparer : IEqualityComparer<string>
```
Comment (ayende) [inline src/Raven.Client/Util/NewLineInsensitiveStringComparer.cs]: I don't like this. This allocates multiple times during the equality process.
We are better off just normalizing the values once in `Compare`

**Review by ayende — CHANGES_REQUESTED** [review]

File: test/SlowTests/Issues/RavenDB_16079.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+using Raven.Client.Documents.Indexes;
+using Raven.Client.Documents.Operations.Indexes;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace SlowTests.Issues
+{
+    public class RavenDB_16079 : RavenTestBase
+    {
+        public RavenDB_16079(ITestOutputHelper output) : base(output)
```
Comment (ml054) [inline test/SlowTests/Issues/RavenDB_16079.cs:11]: I wonder if Git normalizes new line endings during checkout process, so we test X vs X here? Maybe we should replace new line endings on the fly during test execution and then compare?

Comment (ml054) [reply test/SlowTests/Issues/RavenDB_16079.cs:11]: https://github.com/ravendb/ravendb/blob/v5.1/.gitattributes#L5

File: src/Raven.Client/Util/NewLineInsensitiveStringComparer.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Client/Util/NewLineInsensitiveStringComparer.cs]: Simplified

File: test/SlowTests/Issues/RavenDB_16079.cs (hunk 2 above)
Comment (ppekrol) [reply test/SlowTests/Issues/RavenDB_16079.cs:11]: I think that it only normalizes line endings at EOL not in the content?

File: src/Raven.Client/Documents/Indexes/IndexDefinition.cs [hunk 3]
```diff
@@ -17,6 +18,8 @@ namespace Raven.Client.Documents.Indexes
     /// </summary>
     public class IndexDefinition
     {
+        private static readonly Regex _regex = new Regex("\r*\n", RegexOptions.Compiled);
```
Comment (ayende) [inline src/Raven.Client/Documents/Indexes/IndexDefinition.cs]:
```suggestion
        private static readonly Regex _regex = new Regex("\r?\n", RegexOptions.Compiled);
```

That is a more correct way to do that, no?

Comment (ppekrol) [reply src/Raven.Client/Documents/Indexes/IndexDefinition.cs]: In theory yes, in practice... maybe :) Pushed.

File: test/SlowTests/Issues/RavenDB_16079.cs (hunk 2 above)
Comment (ml054) [reply test/SlowTests/Issues/RavenDB_16079.cs:11]: oh you are right - we use " and \r \n inside stings. Ignore my comment then.

**Comment by grisha-kotler** [comment]
Which index will be saved in the database record?
Old one or new one?

**Comment by ml054** [comment]
I guess old one as new one with equal - no there is "no change" ?

**Comment by ppekrol** [comment]
Correct

**Pull request merged**